File: template-api/app/graphql/mutations/message_create.rb
# frozen_string_literal: true

module Mutations
  class MessageCreate < GraphQL::Schema::Mutation
    argument :content, String, required: true
    argument :room_id, String, required: true

    field :message, Types::MessageType, null: true
    field :errors, [String], null: false

    def resolve(content:, room_id:)
      Rails.logger.info "🚀 MessageCreate called with content: '#{content}', room_id: '#{room_id}'"
      current_user = context[:current_user] || context[:user_context]&.current_user
      Rails.logger.info "👤 Current user in mutation: #{current_user&.id}"

      if current_user.nil?
        Rails.logger.warn "❌ No authenticated user found in message create mutation"
        return {
          message: nil,
          errors: ['You must be signed in to send messages']
        }
      end

      message = Message.new(
        content: content,
        room_id: room_id,
        user: current_user
      )

      if message.save
        Rails.logger.info "Message saved: #{message.inspect}"
        
        # Reload the message with associations to ensure proper serialization
        message = Message.includes(:user).find(message.id)
        Rails.logger.info "Message reloaded with user: #{message.user&.name} (ID: #{message.user&.id})"
        
        # Broadcast message via ActionCable using the service
        BroadcastService.broadcast_message(room_id, message)

        {
          message: message,
          errors: []
        }
      else
        {
          message: nil,
          errors: message.errors.full_messages
        }
      end
    end
  end
end

File: template_phoenix_api/lib/template_phoenix_api_web/schema/mutations/message_create.ex
defmodule TemplatePhoenixApiWeb.Schema.Mutations.MessageCreate do
  use Absinthe.Schema.Notation
  
  alias TemplatePhoenixApi.Content
  
  require Logger

  object :message_create_mutation do
    @desc "Create a new chat message"
    field :message_create, :message_create_response do
      arg :content, non_null(:string)
      arg :room_id, non_null(:string)
      
      resolve fn %{content: content, room_id: room_id}, %{context: %{current_user: current_user}} ->
        Logger.info("🚀 GraphQL MessageCreate called with content: '#{content}', room_id: '#{room_id}'")
        Logger.info("👤 Current user in mutation: #{current_user.id}")
        
        message_attrs = %{
          content: content,
          room_id: room_id,
          user_id: current_user.id
        }
        
        case Content.create_message(message_attrs) do
          {:ok, message} ->
            Logger.info("✅ Message saved: #{message.id}")
            
            # Reload the message with user association
            message = Content.get_message!(message.id)
            Logger.info("📄 Message reloaded with user: #{message.user.name} (ID: #{message.user.id})")
            
            # Broadcast message via Phoenix channels using BroadcastService
            TemplatePhoenixApi.Services.BroadcastService.broadcast_message(room_id, message)
            
            {:ok, %{
              message: message,
              errors: []
            }}
            
          {:error, changeset} ->
            Logger.error("❌ Failed to create message: #{inspect(changeset.errors)}")
            errors = format_changeset_errors_list(changeset)
            {:ok, %{
              message: nil,
              errors: errors
            }}
        end
      end
      
      resolve fn _args, _context ->
        {:error, "You must be signed in to send messages"}
      end
    end
  end

  defp format_changeset_errors_list(changeset) do
    changeset.errors
    |> Enum.map(fn {field, {message, _}} ->
      "#{field}: #{message}"
    end)
  end
end